padl/transforms.py

```diff
@@ -710,6 +750,9 @@ def _pd_itercall(self, args, mode: Mode, loader_kwargs: Optional[dict] = None,
         """
         assert mode in ('eval', 'train'), '_pd_itercall can only be used with mode eval or train'
 
+        global _pd_trace
+        _pd_trace.clear()
```

Comment:
I'm not too familiar with how globals in python work, but is this the right place to declare `_pd_trace`? This is also needed for `infer_apply` which doesn't call `_pd_itercall`.

Reply:
`_pd_trace` is also defined on the line 51

Reply:
Ahh I see. Is there a reason you declare it as `global` here, but not on line 51?

Reply:
As I understand the `global` keyword is for changing the value of a global variable inside a function, but I might not know everything about it

Reply:
Okay I see. Do you need to also add the `global _pd_trace` to `infer_apply` as well then?

Reply:
yes I think so

Reply:
Actually for the case of lists it is not needed to use the `global` keyword, because the `append` method adds an element on the memory associated to a list. This is different to what happens if you use a global `integer`, for example, and change its value inside the function.

For example:
```
a = 5

def something():
    a = 0
    print(f'a is {a}`)

something()
print(f' a is {a}')
```
prints:
```
a is 0
a is 5
```
but 
```
a = 5

def something():
    global a
    a = 0
    print(f'a is {a}`)

something()
print(f' a is {a}')
```
prints:
```
a is 0
a is 0
```
if the flobal variable is a list and inside `something` you append an element it is going to be added to the global variable no matter the `global` keyword` is called or not.

Reply:
ok, then removed!